src/TBuffer.cpp

```diff
@@ -2441,75 +2441,113 @@ inline int TBuffer::wrap(int startLine)
     if (static_cast<int>(buffer.size()) < startLine || startLine < 0) {
         return 0;
     }
+
     std::queue<std::deque<TChar>> queue;
     QStringList tempList;
     QStringList timeList;
     QList<bool> promptList;
     int lineCount = 0;
+    const QString lineBreaks = qsl(",.- ，。　()");
+
+    const auto hostFont = mpHost->getDisplayFont();
+    int mFontWidth = QFontMetrics(hostFont).averageCharWidth();
+    int wrapByPixel = mWrapAt * mFontWidth;
+    
+    // Loop from current line up to the last line sitting in buffer
     for (int i = startLine, total = static_cast<int>(buffer.size()); i < total; ++i) {
+        // Capture isPrompt
         bool isPrompt = promptBuffer[i];
         std::deque<TChar> newLine;
         QString lineText = "";
         QString time = timeBuffer[i];
-        int indent = 0;
-        if (static_cast<int>(buffer[i].size()) >= mWrapAt) {
-            for (int i3 = 0; i3 < mWrapIndent; ++i3) {
-                TChar pSpace;
-                newLine.push_back(pSpace);
-                lineText.append(" ");
-            }
-            indent = mWrapIndent;
-        }
-        int lastSpace = 0;
+        // Index where wrap should happen
         int wrapPos = 0;
+        // Number of character in this line
         int length = buffer[i].size();
+
+        // If number of character is zero, this is just blank line, which actually means the for loop after the if statement won't happen
         if (length == 0) {
             tempList.append(QString());
             std::deque<TChar> emptyLine;
             queue.push(emptyLine);
             timeList.append(time);
+            promptList.append(false);
         }
-        for (int i2 = 0, total = static_cast<int>(buffer[i].size()); i2 < total;) {
-            if (length - i2 > mWrapAt - indent) {
-                wrapPos = calculateWrapPosition(i, i2, i2 + mWrapAt - indent);
-                lastSpace = qMax(0, wrapPos);
-            } else {
-                lastSpace = 0;
-            }
-            int wrapPosition = (lastSpace) ? lastSpace : (mWrapAt - indent);
-            for (int i3 = 0; i3 < wrapPosition; ++i3) {
-                if (lastSpace > 0) {
-                    if (i2 > lastSpace) {
-                        break;
-                    }
-                }
-                if (i2 >= static_cast<int>(buffer[i].size())) {
-                    break;
-                }
-                if (lineBuffer[i].at(i2) == '\n') {
-                    i2++;
-                    break;
-                }
+
+        // Track whether there is data that needs to be appended
+        bool hasContent = false;
+
+        // Track the current width in pixel
+        int lineWidth = QFontMetrics(hostFont).horizontalAdvance(lineBuffer[i]);
+        if (lineWidth >= wrapByPixel) {
+            // This is where actual wrapping occurs
+            for (int i2 = 0, total = static_cast<int>(buffer[i].size()); i2 < total;) {
```

Comment:
👎  No. no, no!

This adding a `QChar`/`TChar` at a time to `lineText`/`newLine` is no longer safe to use once we are (and we *are* **now**) handling non-ASCII text. `QString`s contain effectively UTF-16BE encoded text and UTF-16 like UTF-8 is a ***variable length encoding*** system. This means that a single `QChar` may NOT be a single grapheme (what some people think is a "character"), even a pair of them (a high-surrogate followed by a low-surrogate) may not be a grapheme. To step through the "characters" you must step through them "grapheme" by "grapheme" - which means putting the whole line of text (in `lineBuffer.at(i)`) and using `QTextBoundaryFinder` to work out which `QChars` are in each grapheme...

Reply:
You might like to take a look at `(int) TTextEdit::convertMouseXToBufferX(const int mouseX, const int lineNumber, bool* isOutOfbounds, bool* isOverTimeStamp) const` as that does do the right thing as it tries to work out where the mouse is positioned on a line of text on the screen and relate that to the matching horizontal position of the start of the grapheme in the `TBuffer::lineBuffer`.

Reply:
We are very passionate about graphemes here. Thanks @hhsiao for working on this 😬

Reply:
I've made changes to do wrapping base on QTextBoundary, which I believe would prevent my code from cutting character off at the wrong place

Reply:
😊 Sorry @hhsiao if I came across a bit strong in https://github.com/Mudlet/Mudlet/pull/6207#discussion_r935999160 - TBH this was something that I was planning to work on during my foreign holiday/Summer vacation (starting in a couple of weeks) but you have dived in ahead of me and I wanted to steer you away from considering individual `QChar`s as we have been doing...

I have some thoughts on how we might improve things which I'll raise in a thread in the Discord #mudlet-development channel...